## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
#This code was influenced by Ask BCS Tutor, ChatGPT, Xpert Learning Assistance, Pandas Documentation and 3Worlds Website.
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
print("Column Names:")
print(df.columns)


Column Names:
Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics
print("\nBasic Statistics:")
print(df.describe())



Basic Statistics:
          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.



In [5]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
print("\nTop three item categories with the most entries:")
print(top_categories)



Top three item categories with the most entries:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [6]:
# For the category with the most entries,
# which subcategory had the most entries?
category_with_most_entries = top_categories.index[0]
subcategory_most_entries = df[df['category'] == category_with_most_entries]['subcategory'].value_counts().idxmax()
print(f"\nSubcategory with the most entries in the category '{category_with_most_entries}': {subcategory_most_entries}")



Subcategory with the most entries in the category 'consumables': bathroom supplies


In [7]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
print("\nTop five clients with the most entries:")
print(top_clients)



Top five clients with the most entries:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [8]:
# Store the client ids of those top 5 clients in a list.
top_clients_ids = top_clients.index.tolist()
print(f"\nTop 5 clients' IDs: {top_clients_ids}")




Top 5 clients' IDs: [33615, 66037, 46820, 38378, 24741]


In [9]:
# How many total units (the qty column) did the
# client with the most entries order order?
client_with_most_entries = top_clients.index[0]
total_units_ordered = df[df['client_id'] == client_with_most_entries]['qty'].sum()
print(f"\nTotal units ordered by the client with the most entries (client_id={client_with_most_entries}): {total_units_ordered}")




Total units ordered by the client with the most entries (client_id=33615): 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [18]:
#DataFrame Data for Unit_Price and Qty
 

In [19]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']
print(df[['unit_price', 'qty', 'subtotal']])




       unit_price   qty   subtotal
0         1096.80   105  115164.00
1           24.95    21     523.95
2           13.52    39     527.28
3           36.42    29    1056.18
4          195.10    20    3902.00
...           ...   ...        ...
54634       83.13    33    2743.29
54635      206.59    47    9709.73
54636       65.66   475   31188.50
54637        1.48   112     165.76
54638        3.01  1031    3103.31

[54639 rows x 3 columns]


In [42]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
import numpy as np 
df ["total_weight"] = df['unit_weight'] * df['qty']  
shipping_price = np.where(df['total_weight'] > 50, df['total_weight'] * 7, df['total_weight'] * 10) 
df['shipping_price'] = shipping_price
print(df[['total_weight','qty', 'shipping_price']])




       total_weight   qty  shipping_price
0            787.50   105         5512.50
1             31.29    21          312.90
2             65.52    39          458.64
3             35.67    29          356.70
4            928.60    20         6500.20
...             ...   ...             ...
54634         74.25    33          519.75
54635        549.90    47         3849.30
54636       1976.00   475        13832.00
54637       2020.48   112        14143.36
54638       2134.17  1031        14939.19

[54639 rows x 3 columns]


In [43]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
sales_tax_rate = 0.0925
df['total_price'] = round((df['subtotal'] + df['shipping_price']) * (1 + sales_tax_rate), 2)   
print(df[['subtotal', 'shipping_price', 'total_price']])



        subtotal  shipping_price  total_price
0      115164.00         5512.50    131839.08
1         523.95          312.90       914.26
2         527.28          458.64      1077.12
3        1056.18          356.70      1543.57
4        3902.00         6500.20     11364.40
...          ...             ...          ...
54634    2743.29          519.75      3564.87
54635    9709.73         3849.30     14813.24
54636   31188.50        13832.00     49184.90
54637     165.76        14143.36     15632.71
54638    3103.31        14939.19     19711.43

[54639 rows x 3 columns]


In [44]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
print(df[['unit_cost', 'qty', 'shipping_price', 'line_cost']])



       unit_cost   qty  shipping_price  line_cost
0         762.71   105         5512.50   85597.05
1          15.09    21          312.90     629.79
2           7.86    39          458.64     765.18
3          24.85    29          356.70    1077.35
4         108.17    20         6500.20    8663.60
...          ...   ...             ...        ...
54634      51.60    33          519.75    2222.55
54635     175.46    47         3849.30   12095.92
54636      57.31   475        13832.00   41054.25
54637       1.22   112        14143.36   14280.00
54638       1.61  1031        14939.19   16599.10

[54639 rows x 4 columns]


In [45]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['line_cost']
print(df[['line_cost', 'total_price', 'profit']])


       line_cost  total_price    profit
0       85597.05    131839.08  46242.03
1         629.79       914.26    284.47
2         765.18      1077.12    311.94
3        1077.35      1543.57    466.22
4        8663.60     11364.40   2700.80
...          ...          ...       ...
54634    2222.55      3564.87   1342.32
54635   12095.92     14813.24   2717.32
54636   41054.25     49184.90   8130.65
54637   14280.00     15632.71   1352.71
54638   16599.10     19711.43   3112.33

[54639 rows x 3 columns]


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [46]:
# Check your work using the totals above

# Sum the 'total_price' for each order
order_totals = df.groupby('order_id')['total_price'].sum()

# Provided total prices for comparison
provided_totals = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

# Compare the calculated totals with the provided totals
for order_id, provided_total in provided_totals.items():
    calculated_total = order_totals.loc[order_id]
    
    if provided_total is not None:
        print(f"Order ID {order_id}:")
        print(f"  Calculated Total: {calculated_total}")
        print(f"  Provided Total:   {provided_total}")
        print(f"  Match:            {calculated_total == provided_total}")
        print()



Order ID 2742071:
  Calculated Total: 152811.88999999998
  Provided Total:   152811.89
  Match:            False

Order ID 2173913:
  Calculated Total: 162388.71
  Provided Total:   162388.71
  Match:            True

Order ID 6128929:
  Calculated Total: 923441.25
  Provided Total:   923441.25
  Match:            True



## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [47]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

# Identify the top clients by quantity
top_clients_by_quantity = df['client_id'].value_counts().nlargest(5).index

# Filter the DataFrame for the top clients
top_clients_df = df[df['client_id'].isin(top_clients_by_quantity)]

# Calculate the total spending for each of the top clients
top_clients_spending = top_clients_df.groupby('client_id')['total_price'].sum()

# Display the top clients by quantity
print("Top Clients by Quantity:")
print(top_clients_by_quantity)

# Display the total spending for each of the top clients
print("\nTotal Spending for Top Clients:")
print(top_clients_spending)




Top Clients by Quantity:
Index([33615, 66037, 46820, 38378, 24741], dtype='int64', name='client_id')

Total Spending for Top Clients:
client_id
24741    82268892.02
33615     8377308.52
38378    12906550.87
46820     9743794.36
66037    10259514.79
Name: total_price, dtype: float64


In [50]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# Identify the top clients by profit
top_clients_by_profit = df.groupby('client_id')['profit'].sum().nlargest(5).index

# Filter the DataFrame for the top clients
top_clients_df = df[df['client_id'].isin(top_clients_by_profit)]

# Create a summary DataFrame
summary_df = pd.DataFrame({
    'Total Units Purchased': top_clients_df.groupby('client_id')['qty'].sum(),
    'Total Shipping Price': top_clients_df.groupby('client_id')['shipping_price'].sum(),
    'Total Revenue': top_clients_df.groupby('client_id')['total_price'].sum(),
    'Total Profit': top_clients_df.groupby('client_id')['profit'].sum()
})

# Sort the summary DataFrame by total profit
summary_df = summary_df.sort_values(by='Total Profit', ascending=False)

# Display the summary DataFrame
print(summary_df)





           Total Units Purchased  Total Shipping Price  Total Revenue  \
client_id                                                               
78965                    4033302            5340756.88   6.559237e+08   
85931                     644423           10122549.19   2.374170e+08   
38473                     402770            4391003.73   1.835526e+08   
86721                    1010426           64335727.72   1.812663e+08   
24741                     239862            5126448.37   8.226889e+07   

           Total Profit  
client_id                
78965      2.473528e+08  
85931      1.092941e+08  
38473      4.335173e+07  
86721      4.297470e+07  
24741      3.657999e+07  


In [51]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

# Identify the top clients by profit
top_clients_by_profit = df.groupby('client_id')['profit'].sum().nlargest(5).index

# Filter the DataFrame for the top clients
top_clients_df = df[df['client_id'].isin(top_clients_by_profit)]

# Create a summary DataFrame in millions of dollars
summary_df = pd.DataFrame({
    'Total Units Purchased (Millions)': top_clients_df.groupby('client_id')['qty'].sum() / 1_000_000,
    'Total Shipping Price (Millions)': top_clients_df.groupby('client_id')['shipping_price'].sum() / 1_000_000,
    'Total Revenue (Millions)': top_clients_df.groupby('client_id')['total_price'].sum() / 1_000_000,
    'Total Profit (Millions)': top_clients_df.groupby('client_id')['profit'].sum() / 1_000_000
})

# Sort the summary DataFrame by total profit
summary_df = summary_df.sort_values(by='Total Profit (Millions)', ascending=False)

# Display the summary DataFrame
print(summary_df)



           Total Units Purchased (Millions)  Total Shipping Price (Millions)  \
client_id                                                                      
78965                              4.033302                         5.340757   
85931                              0.644423                        10.122549   
38473                              0.402770                         4.391004   
86721                              1.010426                        64.335728   
24741                              0.239862                         5.126448   

           Total Revenue (Millions)  Total Profit (Millions)  
client_id                                                     
78965                    655.923670               247.352842  
85931                    237.417001               109.294129  
38473                    183.552572                43.351727  
86721                    181.266297                42.974695  
24741                     82.268892                36.579992 

In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
# Done in previous cell


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M

In [52]:
# Identify the top clients by profit
top_clients_by_profit = df.groupby('client_id')['profit'].sum().nlargest(5).index

# Filter the DataFrame for the top clients
top_clients_df = df[df['client_id'].isin(top_clients_by_profit)]

# Create a summary DataFrame in millions of dollars
summary_df = pd.DataFrame({
    'Total Units Purchased (Millions)': top_clients_df.groupby('client_id')['qty'].sum() / 1_000_000,
    'Total Shipping Price (Millions)': top_clients_df.groupby('client_id')['shipping_price'].sum() / 1_000_000,
    'Total Revenue (Millions)': top_clients_df.groupby('client_id')['total_price'].sum() / 1_000_000,
    'Total Profit (Millions)': top_clients_df.groupby('client_id')['profit'].sum() / 1_000_000
})

# Sort the summary DataFrame by total profit from highest to lowest
summary_df = summary_df.sort_values(by='Total Profit (Millions)', ascending=False)

# Display the sorted summary DataFrame
print(summary_df)


           Total Units Purchased (Millions)  Total Shipping Price (Millions)  \
client_id                                                                      
78965                              4.033302                         5.340757   
85931                              0.644423                        10.122549   
38473                              0.402770                         4.391004   
86721                              1.010426                        64.335728   
24741                              0.239862                         5.126448   

           Total Revenue (Millions)  Total Profit (Millions)  
client_id                                                     
78965                    655.923670               247.352842  
85931                    237.417001               109.294129  
38473                    183.552572                43.351727  
86721                    181.266297                42.974695  
24741                     82.268892                36.579992 